lsp-methods.el

```diff
@@ -521,10 +521,21 @@ directory."
 
       (puthash root lsp--cur-workspace lsp--workspaces)
       (run-hooks 'lsp-before-initialize-hook)
-      (setq init-params `(:processId ,(emacs-pid) :rootPath ,root
-                           :rootUri ,(concat "file://" root)
-                           :capabilities ,(lsp--client-capabilities)))
-      (setf response (lsp--send-request (lsp--make-request "initialize" init-params)))
+      (setq init-params
+            (lsp--merge-plists
+             `(:processId ,(emacs-pid)
+                :rootPath ,root
+                :rootUri ,(concat "file://" root)
+                :capabilities ,(lsp--client-capabilities))
+             (if extra-init-params
```

Comment:
What about this:
```emacs-lisp
`(:processId ,(emacs-pid)
 :rootpath ,root
 :rootUri ,(concat "file://" root)
 :capabilities ,(lsp--client-capabilities)
 :initializationOptions ,(if (functionp extra-init-params) (funcall extra-init-params lsp--cur-workspace) extra-init-params))
```
If `extra-init-params` is nil, `json-encode` will marshal it to `null`.

Reply:
Yeah, this is a lot cleaner. I'll fixup the commit and push.

(I wanted to avoid even including the `initializationOptions` JSON blob if it wasn't specified in the client, but I guess if it's in the protocol clients should at least be able to ignore the unexpected field.)